library/classes.lisp

```diff
@@ -110,8 +67,28 @@
     (Ok :good)
     (Err :bad))
 
+  ;;
+  ;; Eq
+  ;;
+
+  (define-class (Eq :a)
+    "Types which have equality defined."
+    (== (:a -> :a -> Boolean)))
+
+  (define-instance (Eq types:LispType)
+    (define (== a b)
+      (lisp Boolean (a b)
+        (cl:equalp a b))))
+
+  (define-class (Eq :a => Num :a)
+    "Types which have numeric operations defined."
+    (+ (:a -> :a -> :a))
+    (- (:a -> :a -> :a))
+    (* (:a -> :a -> :a))
+    (fromInt (Integer -> :a)))
+ 
   (define-instance (Eq Unit)
-    (define (== _a _b) True))
+    (define (== _ _) True))
```

Comment:
Avoiding a type check at runtime.

Reply:
Could we imagine eliminating this type check as a compiler optimization in the future?

Reply:
Yeah probably. ADTs aren't sealed because you can always recompile the define-type form with another constructor. We could either special case stdlib types, or assume that ADTs are sealed in release mode.

Reply:
I vote for assuming ADTs are sealed in release mode. Redefinition in release mode is already janky since we emit `defstruct` instead of `defclass`, and a lot of compelling optimizations will depend on redefinition being UB.